### PR TITLE
build head~1 and run pr-time benchmarks on it

### PR DESCRIPTION
**Pull request by laithsakka**
This PR will checkout the commit HEAD~1 and build it and then run the benchmarks on it and store the results 
in a file. Next PR will add the comparison of numbers and the regression reporting. 
check results by checking logs o the test (https://github.com/pytorch/pytorch/actions/runs/10149355968/job/28065140121)
```
benchmark results on main: 
+ cat /var/lib/jenkins/workspace/test/test-reports/pr_time_benchmarks_before.txt
update_hint_regression,instruction_count,29199009373



benchmark results on current PR: 
+ cat /var/lib/jenkins/workspace/test/test-reports/pr_time_benchmarks_after.txt
update_hint_regression,instruction_count,28556649567

```
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131751

### CODE REVIEW

File: .ci/pytorch/test.sh [hunk 1]
```diff
@@ -400,7 +400,15 @@ pr_time_benchmarks() {
   source benchmarks/dynamo/pr_time_benchmarks/benchmark_runner.sh "$TEST_REPORTS_DIR/pr_time_benchmarks_after.txt" "benchmarks/dynamo/pr_time_benchmarks/benchmarks"
   echo "benchmark results on current PR: "
   cat  "$TEST_REPORTS_DIR/pr_time_benchmarks_after.txt"
-
+  git checkout HEAD~1
+
+  # we can try firt to run develop alone, and if it worked then we do not need to run clean->develop.
+  # another suggestion was to use historical stored data insetad of building Head~1
+  python setup.py clean
+  python setup.py develop
```
Comment (aorenste) [inline .ci/pytorch/test.sh:408]: Maybe
```
python setup.py develop --cmake
```
which should be good enough for 99% of changes and is a lot faster than clean.

File: .ci/pytorch/test.sh [hunk 2]
```diff
@@ -400,7 +400,15 @@ pr_time_benchmarks() {
   source benchmarks/dynamo/pr_time_benchmarks/benchmark_runner.sh "$TEST_REPORTS_DIR/pr_time_benchmarks_after.txt" "benchmarks/dynamo/pr_time_benchmarks/benchmarks"
   echo "benchmark results on current PR: "
   cat  "$TEST_REPORTS_DIR/pr_time_benchmarks_after.txt"
-
+  git checkout HEAD~1
```
Comment (aorenste) [inline .ci/pytorch/test.sh:403]: Do we need to worry about the fact that we're leaving it checked out at HEAD~1?  Maybe restore it before exiting the function.

File: .ci/pytorch/test.sh [hunk 3]
```diff
@@ -400,7 +400,15 @@ pr_time_benchmarks() {
...
+
+  # we can try firt to run develop alone, and if it worked then we do not need to run clean->develop.
```
Comment (aorenste) [inline .ci/pytorch/test.sh:405]: nit: spelling

**Comment by kit1980** [comment]
`test_forward_backward_compatibility` in the test.sh does basically what you want - compares with the base commit.
You should probably implement yours the same way unless there are serious reasons that will not work for you.

File: .ci/pytorch/test.sh (hunk 2 above)
Comment (kit1980) [reply .ci/pytorch/test.sh:403]: What if there are more than 1 commit in the PR?
You should look at `test_forward_backward_compatibility` and probably do the same.

Comment (laithsakka) [reply .ci/pytorch/test.sh:403]: @aorenste I do exit after i run the pr_time_benchmarks tests, so nothing after that run, no need to restore.
@kit1980 I can check and try to rebuild same way as in test_forward_backward_compatibility, maybe i can move that to  a shared function

File: .ci/pytorch/test.sh (hunk 1 above)
Comment (laithsakka) [reply .ci/pytorch/test.sh:408]: and we fail for the remaining 1%? i will try to mimic test_forward_backward_compatibility

**Comment by laithsakka** [comment]
check if linux.gcp.a100 have hardware counters.

**Comment by laithsakka** [comment]
hide build logs,

**Review by ezyang — CHANGES_REQUESTED** [review]
No. I am seriously skeptical of this approach. There are several reasons why I do not want to do it this way:

1. It takes a long time. If I am iterating on compile time on a fixed revision, there is no reason to keep rebuilding the HEAD~ commit over and over again. But I will have to wait for that on every PR. Building PyTorch is expensive!
2. To make matters worse, you're building PyTorch on the test runner. If it's a GPU runner, you're wasting precious GPU cycles doing a CPU only build.
3. If the only signal you are getting is pairwise performance signals between HEAD~1 and HEAD, then if someone ignores this signal and lands their PR anyway, then it is completely lost to the ages. We absolutely need some alternate dashboard like setup to durably observe regressions on main when people ignore CI signals and land anyway.

I am pretty sure we need a database of some sort. If you think you're not the right person to implement a database, let's find someone else to do it. But definitely not this.

**Comment by oulgen** [comment]
Regarding the time it takes to build, most PRs only change python files, so if we can setup it up right, the second build should be free?

**Comment by laithsakka** [comment]
> No. I am seriously skeptical of this approach. There are several reasons why I do not want to do it this way:
> 
> 1. It takes a long time. If I am iterating on compile time on a fixed revision, there is no reason to keep rebuilding the HEAD~ commit over and over again. But I will have to wait for that on every PR. Building PyTorch is expensive!
> 2. To make matters worse, you're building PyTorch on the test runner. If it's a GPU runner, you're wasting precious GPU cycles doing a CPU only build.
> 3. If the only signal you are getting is pairwise performance signals between HEAD~1 and HEAD, then if someone ignores this signal and lands their PR anyway, then it is completely lost to the ages. We absolutely need some alternate dashboard like setup to durably observe regressions on main when people ignore CI signals and land anyway.
> 
> I am pretty sure we need a database of some sort. If you think you're not the right person to implement a database, let's find someone else to do it. But definitely not this.


- We can always upload the result files somewhere and have a delayed signal as well, but we do need a diff time signal for sure, because its not practical to depend on a late signal and revert not is it practical to depend on someone monitering those external delayed dashboard. 
So i think step 1 we need to get the diff time signal working. Then we can think about the next step. 

- We did have the concern about rebuilding pytorch, We discussed three other alternatives :
1. store the most recent result (The result on main) in some storage and compare against it. 
2. install nightly and compare against it. 
3. do not build from scratch but rather, apply change and try to build (this could fail in such case we do not run the test). 
4. we can do this a push time test (when its landed not while iterating). 

"I am pretty sure we need a database of some sort."
if you have ideas please share

 "If you think you're not the right person to implement a database, let's find someone else to do it."
I think lets firs decide what we want to do, then we can decide who is the right person to do it.

**Comment by laithsakka** [comment]
> We used to have a version of this (by @robieta?), maybe he has more thoughts

ahh interesting

**Comment by laithsakka** [comment]
Me too

**Comment by laithsakka** [comment]
Me too

**Comment by laithsakka** [comment]
this comment is auto generatd testing

**Comment by laithsakka** [comment]
this comment is auto generatd testing

**Comment by laithsakka** [comment]
this comment is auto generatd testing ,

**Comment by laithsakka** [comment]
Got good amount of push back from pytorch on two builds idea will pivot to something else and close this for now.

**Pull request closed without merging**